hudi-utilities/src/main/java/org/apache/hudi/utilities/HoodieMetadataTableValidator.java

```diff
@@ -0,0 +1,458 @@
+   */
+  private void validateLatestBaseFiles(HoodieTableFileSystemView metaFsView, HoodieTableFileSystemView fsView, String partitionPath) {
+
+    List<HoodieBaseFile> latestFilesFromMetadata = metaFsView.getLatestBaseFiles(partitionPath).sorted(new HoodieBaseFileCompactor()).collect(Collectors.toList());
+    List<HoodieBaseFile> latestFilesFromFS = fsView.getLatestBaseFiles(partitionPath).sorted(new HoodieBaseFileCompactor()).collect(Collectors.toList());
+
+    LOG.info("Latest base file from metadata: " + latestFilesFromMetadata);
+    LOG.info("Latest base file from direct listing: " + latestFilesFromFS);
+    if (!latestFilesFromMetadata.equals(latestFilesFromFS)) {
+      String message = "Validation of metadata get latest base file for partition " + partitionPath + " failed. "
+          + "Latest base file from metadata: " + latestFilesFromMetadata
+          + "Latest base file from direct listing: " + latestFilesFromFS;
+      LOG.error(message);
+      throw new HoodieValidationException(message);
+    } else {
+      LOG.info("Validation of getLatestBaseFiles success.");
+    }
+  }
+
+  /**
+   * Compare getLatestFileSlices between metadata table and fileSystem.
+   */
+  private void validateLatestFileSlices(HoodieTableFileSystemView metaFsView, HoodieTableFileSystemView fsView, String partitionPath) {
+
+    List<FileSlice> latestFileSlicesFromMetadataTable = metaFsView.getLatestFileSlices(partitionPath).sorted(new FileSliceCompactor()).collect(Collectors.toList());
+    List<FileSlice> latestFileSlicesFromFS = fsView.getLatestFileSlices(partitionPath).sorted(new FileSliceCompactor()).collect(Collectors.toList());
+
+    LOG.info("Latest file list from metadata: " + latestFileSlicesFromMetadataTable);
+    LOG.info("Latest file list from direct listing: " + latestFileSlicesFromFS);
+    if (!latestFileSlicesFromMetadataTable.equals(latestFileSlicesFromFS)) {
+      String message = "Validation of metadata get latest file slices for partition " + partitionPath + " failed. "
+          + "Latest file list from metadata: " + latestFileSlicesFromMetadataTable
+          + "Latest file list from direct listing: " + latestFileSlicesFromFS;
+      LOG.error(message);
+      throw new HoodieValidationException(message);
+    } else {
+      LOG.info("Validation of getLatestFileSlices success.");
+    }
+  }
+
+  private HoodieTableFileSystemView createHoodieTableFileSystemView(HoodieSparkEngineContext engineContext, boolean enableMetadataTable) {
+
+    HoodieMetadataConfig metadataConfig = HoodieMetadataConfig.newBuilder()
+        .enable(enableMetadataTable)
+        .withAssumeDatePartitioning(cfg.assumeDatePartitioning)
+        .build();
+
+    return FileSystemViewManager.createInMemoryFileSystemView(engineContext,
+        metaClient, metadataConfig);
+  }
+
+  protected boolean compareCollectionResult(Collection<String> c1, Collection<String> c2) {
+    return c1.size() == c2.size() && c1.containsAll(c2) && c1.containsAll(c2);
```

Comment:
Could `containsAll` be slow, if the Collection is List?

Reply:
Remove and use sort and .equals to compare.